### PR TITLE
Jax backend

### PR DESCRIPTION
**Pull request by eickenberg**
Here is the jax backend for 1D. Tests pass.

I will add 2D and 3D here in the next days.

Please start reviewing this when you get a chance.

### CODE REVIEW

File: tests/scattering1d/test_jax_backend_1d.py
Comment (janden) [inline tests/scattering1d/test_jax_backend_1d.py:36]: EOL

File: tests/general/test_jnp_vs_np.py
Comment (janden) [inline tests/general/test_jnp_vs_np.py:26]: Remove these intro comments in each method.

File: tests/general/test_jnp_vs_np.py
Comment (janden) [inline tests/general/test_jnp_vs_np.py:1]: It would be good to have a brief description of this test since it doesn't fit in with the others. In some sense, it's a test of Jax, correct?

File: tests/general/test_jnp_vs_np.py
Comment (janden) [inline tests/general/test_jnp_vs_np.py:40]: How is this different from the previous test?

File: tests/scattering1d/test_jax_backend_1d.py
Comment (janden) [inline tests/scattering1d/test_jax_backend_1d.py:1]: Rename `test_jax_scattering1d.py` per convention.

File: tests/scattering1d/test_jax_backend_1d.py
Comment (janden) [inline tests/scattering1d/test_jax_backend_1d.py:34]: Do we know that Jax is actually being used here (as opposed to NP)? There should be a way to find out using the mock package, I think. Conversely, it would be good to make sure that NP *is not* being used (except for filter generation) so that a call doesn't slip through by accident.

**Review by janden — CHANGES_REQUESTED** [review]
Great work! Some things do discuss here. Mainly clarifying the triple inheritance stuff which scare me a bit.

File: kymatio/scattering1d/backend/jax_backend.py
Comment (eickenberg) [reply kymatio/scattering1d/backend/jax_backend.py:5]: That is a good point.

Here is some pictorial evidence that it works:

image

I'll leave this comment open because I haven't changed this yet, but probably will to enhance readability.

File: kymatio/scattering1d/frontend/jax_frontend.py
Comment (eickenberg) [reply kymatio/scattering1d/frontend/jax_frontend.py:5]: Explanation:
1. Multiple inheritance generally looks for methods/attributes in the order of the classes specified above. Hence `ScatteringJax` goes first in order to have priority adding its attributes in.
2. Apart from the inheritance of `ScatteringNumpPy1D` this class and its `__init__` look exactly like the one in `ScatteringNumPy1D` itself.
3. I had to include `ScatteringBase1D` because I needed to overwrite the `__init__`, basically copying and pasting the `__init__` from `ScatteringNumPy1D` but with the backend changed. I had to do that because otherwise it was doing something numpy-specific from the `ScatteringNumPy1D.__init__`. In order to do that overwriting I needed access to `ScatteringBase1D` (from which `ScatteringNumPy1D` also inherits and which it uses in its `__init__`)
4. I initially tried switching positions 2 and 3 and I got a multiple inheritance error (because `ScatteringNumPy1D` derives from `ScatteringBase1D`) and so not all permutations are actually allowed.

Will add some comments

File: tests/general/test_jnp_vs_np.py
Comment (eickenberg) [reply tests/general/test_jnp_vs_np.py:1]: This is correct. I'll add a comment. It's just an installment to check the functionality we need for correspondence up to a certain precision with numpy

File: tests/general/test_jnp_vs_np.py
Comment (eickenberg) [reply tests/general/test_jnp_vs_np.py:26]: ah lol yeah

File: tests/general/test_jnp_vs_np.py
Comment (eickenberg) [reply tests/general/test_jnp_vs_np.py:40]: don't know! I copied and pasted from the previous PR :D Will check

File: tests/scattering1d/test_jax_backend_1d.py
Comment (eickenberg) [reply tests/scattering1d/test_jax_backend_1d.py:1]: thanks!

File: tests/scattering1d/test_jax_backend_1d.py
Comment (eickenberg) [reply tests/scattering1d/test_jax_backend_1d.py:34]: image

Here is another pictorial proof.

You mean build a fake numpy or something that errors on all/most methods? With https://docs.python.org/dev/library/unittest.mock.html ?

File: tests/scattering1d/test_jax_backend_1d.py
Comment (eickenberg) [reply tests/scattering1d/test_jax_backend_1d.py:36]: thanks!

File: tests/scattering1d/test_jax_backend_1d.py
Comment (eickenberg) [inline tests/scattering1d/test_jax_backend_1d.py:23]: huh, why were we doing this? Instead of `np.load`?

**Review by eickenberg — COMMENTED** [review]
Thanks for the review!

I left some comments about your questions.

Going to address them code-side now

File: kymatio/scattering1d/backend/jax_backend.py
Comment (eickenberg) [reply kymatio/scattering1d/backend/jax_backend.py:5]: I was about to comment on how I only added a comment because I thought that adding `_np` and `_fft` manually would be confusing because of it already being in `JaxBackend`. But your suggestion is to remove `JaxBackend`.

I see. That is definitely worth considering.

File: kymatio/scattering1d/frontend/jax_frontend.py
Comment (eickenberg) [reply kymatio/scattering1d/frontend/jax_frontend.py:5]: added some comments

File: tests/scattering1d/test_jax_scattering1d.py
Comment (eickenberg) [inline tests/scattering1d/test_jax_scattering1d.py:38]: added this extra check here.
previous comments are gone because of filename change

**Comment by eickenberg** [comment]
Added 2D jax backend with passing tests. This required moving some of the numpy input checks from the scattering function to the numpy backend

**Comment by eickenberg** [comment]
Added 3D backend.

Another quick review and discussion of the inheritance situation would be great!

File: kymatio/scattering1d/backend/jax_backend.py
Comment (janden) [reply kymatio/scattering1d/backend/jax_backend.py:5]: That is correct. I'll leave it up to you to decide.

File: kymatio/scattering1d/frontend/jax_frontend.py
Comment (janden) [reply kymatio/scattering1d/frontend/jax_frontend.py:5]: For 3. I don't know that you need to include `ScatteringBase1D` in the inheritance list just to call its constructor. It is a super-super-class of the current one, so there's no issue calling its constructor like you're doing. I tried removing this locally and tests pass.

File: tests/general/test_jnp_vs_np.py
Comment (janden) [reply tests/general/test_jnp_vs_np.py:7]: Yes we need this as a canary of sorts.

File: tests/scattering1d/test_jax_backend_1d.py
Comment (janden) [reply tests/scattering1d/test_jax_backend_1d.py]: > You mean build a fake numpy or something that errors on all/most methods?

Yeah or is that overkill, do you think? You can just do things like `.called` on various methods. Maybe NP will crash already if you give it Jax arrays.

**Comment by janden** [comment]
Ok I'll review 2D and 3D once we've converged on 1D.

**Comment by eickenberg** [comment]
@janden could we touch base on this one soon-ish and try to converge on a design decision?

**Comment by MuawizChaudhary** [comment]
Whats the current state of this PR @janden @eickenberg ? I can jump in and help. Sorry I never finished this during my internship!

**Comment by lostanlen** [comment]
I don't have the relevant background in JAX but i heartily support this: JAX looks really easy to use as a drop-in replacement for NumPy !

**Review by MuawizChaudhary — COMMENTED** [review]
tests error out on the numpy 3d. Why did the input checks have to be changed for numpy again?

**Comment by MuawizChaudhary** [comment]
@janden I fixed up various things in the jax front/backends. So this PR is ready for your review. Unfortunately, it appears that Jenkins does not have Jax installed. So that needs fixing.

**Review by MuawizChaudhary — APPROVED** [review]
I think we could merge this in at any time. Unsure if we wanna do it soon tho, ot later closer to release.

**Comment by lostanlen** [comment]
> DEPRECATION: Python 3.5 reached the end of its life on September 13th, 2020. Please upgrade your Python as Python 3.5 is no longer maintained. pip 21.0 will drop support for Python 3.5 in January 2021. pip 21.0 will remove support for this functionality.

shall we remove py35 from Pip CI ?

**Comment by MuawizChaudhary** [comment]
> > DEPRECATION: Python 3.5 reached the end of its life on September 13th, 2020. Please upgrade your Python as Python 3.5 is no longer maintained. pip 21.0 will drop support for Python 3.5 in January 2021. pip 21.0 will remove support for this functionality.
> 
> shall we remove py35 from Pip CI ?

I vote yes

**Comment by lostanlen** [comment]
I have attempted to update the GitHub worfklow in a recent PR (#892 )
I would say that JAX is next in line after this

**Comment by janden** [comment]
I agree. We'll rebase this once #892 is merged, then get back to reviewing it.

**Comment by MuawizChaudhary** [comment]
@janden i think you have the ability to reset tests for a pr?

**Comment by janden** [comment]
> @janden i think you have the ability to reset tests for a pr?

Yes but it won't make a difference until we've rebased.

**Comment by lostanlen** [comment]
Alright, i'm rebasing. Everyone stand back

File: kymatio/frontend/numpy_frontend.py
Comment (lostanlen) [reply kymatio/frontend/numpy_frontend.py:15]: i'll look into it later

File: kymatio/scattering1d/frontend/jax_frontend.py
Comment (lostanlen) [reply kymatio/scattering1d/frontend/jax_frontend.py]: @MuawizChaudhary update inheritance?

File: kymatio/scattering1d/backend/jax_backend.py
Comment (lostanlen) [reply kymatio/scattering1d/backend/jax_backend.py:5]: @eickenberg make a decision?

**Comment by janden** [comment]
> Jenkins is failing ... Torch 2D ?
> 
> ```python
> tests/scattering2d/test_torch_scattering2d.py ...............sSending interrupt signal to process
> Terminated
> script returned exit code 143
> ```

My guess is that rebuilding the docker image took too much time and there wasn't enough time to run the tests. I could be wrong though. Re-running to check.

**Comment by janden** [comment]
Something happened with the rebase. There are lots of old commits that shouldn't be here. Do you have access to the original tree?

**Comment by janden** [comment]
> Ok looks like I was able to restore it (see #898). Is it ok if I force-push that branch on here so that we can continue the review in the same place?

Ping @lostanlen

**Comment by lostanlen** [comment]
@janden yes, it's all good to me. apologies for the mishap in rebase

File: tests/scattering2d/test_jax_scattering2d.py
Comment (janden) [inline tests/scattering2d/test_jax_scattering2d.py:1]: So there are a number of tests here that don't pertain directly to Jax. Specifically, they end up just running backend primitives (pad, unpad, modulus, etc.) with NP inputs. I haven't checked, but I would bet they're identical to the NP backend tests. So the question is, what are we testing here? That the Jax library properly passes these through to the underlying NP implementation?

Another point is that we don't have the same set of tests for 1D and 3D, so why have them here?

File: tests/scattering2d/test_jax_scattering2d.py
Comment (janden) [inline tests/scattering2d/test_jax_scattering2d.py:163]: Shouldn't we be casting to JNP arrays or doing `device_put` or something here (like in 1D)?

File: tests/scattering2d/test_jax_scattering2d.py
Comment (janden) [inline tests/scattering2d/test_jax_scattering2d.py:186]: Do we need these? Should be enough with the sample data check above?

**Review by janden — CHANGES_REQUESTED** [review]
Looks good. Mostly comments and questions about testing here, which I think can be streamlined. Since the implementation so closely mirrors the NP frontend, I think we can get away with a simple test against sample data for each dimension and be done (perhaps checking on host and on device to be safe). This will ensure that Jax is doing the correct thing. All the other tests basically reproduce the tests we have for the NP frontend.

**Comment by MuawizChaudhary** [comment]
yeah, I'm at CVPR rn but I'll get to it tonight

File: tests/scattering2d/test_jax_scattering2d.py
Comment (MuawizChaudhary) [reply tests/scattering2d/test_jax_scattering2d.py]: we should also be parameterizing Jax with cpu and gpu device in that instance then...

**Comment by MuawizChaudhary** [comment]
... i fucked up

**Comment by janden** [comment]
Cherry-picked your commits onto the cleaned-up branch. Just remember to rebase your branch before you commit and push next time.

File: tests/scattering2d/test_jax_scattering2d.py
Comment (janden) [reply tests/scattering2d/test_jax_scattering2d.py]: I agree.

File: kymatio/frontend/numpy_frontend.py
Comment (eickenberg) [inline kymatio/frontend/numpy_frontend.py:14]: Is this also from another PR or added during this one?

File: kymatio/scattering2d/frontend/numpy_frontend.py
Comment (eickenberg) [inline kymatio/scattering2d/frontend/numpy_frontend.py:17]: again the question whether this numpy file modification is an integral part of this PR

File: kymatio/scattering3d/backend/numpy_backend.py
Comment (eickenberg) [inline kymatio/scattering3d/backend/numpy_backend.py:52]: is this modification to accommodate jax, because jax doesnt allow setting subarrays?
this will likely double the memory usage - worth considering whether to override this only in jax and keep numpy as it was?

File: kymatio/scattering3d/backend/numpy_backend.py
Comment (eickenberg) [inline kymatio/scattering3d/backend/numpy_backend.py:53]: why is the cast to float32 necessary? isn't the input also float32?

File: kymatio/scattering3d/frontend/numpy_frontend.py
Comment (eickenberg) [inline kymatio/scattering3d/frontend/numpy_frontend.py:23]: this is to let jax arrays through as well?
I like the change, just need to understand the motivation. I would have assumed that since in the jax case, `np = jax.numpy`, it should all be transparent

File: tests/scattering2d/test_jax_scattering2d.py
Comment (eickenberg) [reply tests/scattering2d/test_jax_scattering2d.py:1]: has this been addressed? I can't tell

**Review by eickenberg — COMMENTED** [review]
Looks great to me.

I had some questions wherever numpy files were modified, wondering whether that was from some previous PR or whether it pertained to making the structure jax compatible. I don't think any changes need to be made, but I'd like to understand the instances where the numpy files are modified in terms of this question, so comments appreciated.

I would also add an "Approve" to this PR, but it's grayed out for me, because I am the original author of this PR :joy_cat:

**Review by MuawizChaudhary — APPROVED** [review]

**Review by janden — APPROVED** [review]
Great work!

**Review by lostanlen — APPROVED** [review]

**Pull request merged**
